frame/safe-mode/src/lib.rs

```diff
@@ -0,0 +1,522 @@
+	fn do_extend(who: Option<T::AccountId>, duration: T::BlockNumber) -> DispatchResult {
+		if let Some(who) = who {
+			let reserve = T::ExtendReservationAmount::get().ok_or(Error::<T>::NotConfigured)?;
+			Self::reserve(who, reserve)?;
+		}
+
+		let limit =
+			ActiveUntil::<T>::take().ok_or(Error::<T>::IsInactive)?.saturating_add(duration);
+		ActiveUntil::<T>::put(limit);
+		Self::deposit_event(Event::<T>::Extended { block: limit });
+		Ok(())
+	}
+
+	/// Logic for the [`crate::Pallet::force_deactivate`] call.
+	///
+	/// Errors if the safe-mode is already inactive.
+	/// Does not check the origin.
+	fn do_deactivate(reason: ExitReason) -> DispatchResult {
+		let _limit = ActiveUntil::<T>::take().ok_or(Error::<T>::IsInactive)?;
+		Self::deposit_event(Event::Exited { reason });
+		Ok(())
+	}
+
+	/// Logic for the [`crate::Pallet::release_reservation`] and
+	/// [`crate::Pallet::force_release_reservation`] calls.
+	///
+	/// Errors if the safe-mode is active with [`Error::IsActive`].
+	/// Errors if release is called too soon by anyone but [`Config::ForceReservationOrigin`] with
+	/// [`Error::CannotReleaseYet`]. Does not check the origin.
+	fn do_release(
+		who: Option<T::AccountId>,
+		account: T::AccountId,
+		block: T::BlockNumber,
+	) -> DispatchResult {
+		ensure!(!Self::is_active(), Error::<T>::IsActive);
+
+		let reserve = Reservations::<T>::get(&account, &block).ok_or(Error::<T>::NoReservation)?;
+
+		if who.is_some() {
+			let delay = T::ReleaseDelay::get().ok_or(Error::<T>::NotConfigured)?;
+			let now = <frame_system::Pallet<T>>::block_number();
+			ensure!(now > (block.saturating_add(delay)), Error::<T>::CannotReleaseYet);
+		}
+
+		Reservations::<T>::remove(&account, &block);
+		T::Currency::unreserve_named(&block, &account, reserve);
+		Self::deposit_event(Event::<T>::ReservationReleased { block, account, amount: reserve });
+		Ok(())
+	}
+
+	/// Logic for the [`crate::Pallet::slash_reservation`] call.
+	///
+	/// Errors if the safe-mode is active with [`Error::IsActive`].
+	/// Does not check the origin.
+	fn do_slash(account: T::AccountId, block: T::BlockNumber) -> DispatchResult {
+		ensure!(!Self::is_active(), Error::<T>::IsActive);
+		let reserve = Reservations::<T>::take(&account, block).ok_or(Error::<T>::NoReservation)?;
+
+		T::Currency::slash_reserved_named(&block, &account, reserve);
+		Self::deposit_event(Event::<T>::ReservationSlashed { block, account, amount: reserve });
+		Ok(())
+	}
+
+	/// Reserve `reserve` amount from `who` and store it in `Reservations`.
+	fn reserve(who: T::AccountId, reserve: BalanceOf<T>) -> DispatchResult {
+		let block = <frame_system::Pallet<T>>::block_number();
+		T::Currency::reserve_named(&block, &who, reserve)?;
+		let current_reservation = Reservations::<T>::get(&who, block).unwrap_or_default();
+		// Reservation is mapped to the block that an extrinsic calls activate or extend,
+		// therefore we prevent abuse in a block by adding to present value in the same block. TODO:
```

Comment:
I dont quite get what you mean. But it should be possible to call `activate` and `extend` in the same block IMO.